### PR TITLE
Feature/json consumer

### PR DESCRIPTION
**Pull request by tomaszmichalak**
Relates to #134.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/model/FragmentExecutionLog.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package io.knotx.fragments.handler.consumer.html.model;
+package io.knotx.fragments.handler.consumer.api.model;
```
Comment (marcinus) [inline handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/model/FragmentExecutionLog.java]: 👍

File: handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/FragmentExecutionLogConsumer.java [hunk 2]
```diff
@@ -13,26 +13,26 @@
...
+   * Gets a list of processed and unprocessed fragments (execution logs).
    *
-   * @param clientRequest - client request
-   * @param fragmentEvents - all fragment events
-   * @param tasksMetadata - mapping from fragment id to associated task's metadata
+   * @param request - original request data
+   * @param executions - list of fragment execution logs
    */
-  void accept(ClientRequest clientRequest, List<FragmentEvent> fragmentEvents, TasksMetadata tasksMetadata);
+  void accept(ClientRequest request, List<FragmentExecutionLog> executions);
```
Comment (marcinus) [inline handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/FragmentExecutionLogConsumer.java]: Just wondering if this is enough information for any generic consumer we expect to appear. ClientRequest now serves as activation condition for a consumer - could there be any more conditions or alternative usages?
In any case, it really seems to be sufficient at the moment.

File: handler/core/src/main/java/io/knotx/fragments/handler/consumer/FragmentExecutionLogConsumersNotifier.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  private static final Logger LOGGER = LoggerFactory
+      .getLogger(FragmentExecutionLogConsumersNotifier.class);
+
+  private final List<FragmentExecutionLogConsumer> consumers;
+
+  public FragmentExecutionLogConsumersNotifier(List<FactoryOptions> options) {
+    ServiceLoader<FragmentExecutionLogConsumerFactory> loader = ServiceLoader
+        .load(FragmentExecutionLogConsumerFactory.class);
+    this.consumers = options.stream()
+        .map(o -> StreamSupport.stream(loader.spliterator(), false)
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/handler/consumer/FragmentExecutionLogConsumersNotifier.java]: Extracting the operations in `map` to separate method would increase readibility

File: handler/core/src/test/java/io/knotx/fragments/handler/consumer/EventLogConverterTest.java [hunk 4]
```diff
@@ -43,7 +44,7 @@ void fillWithEmptyLog() {
 
     NodeExecutionData result = tested.getExecutionData(NODE_ID);
 
-    assertEquals(LoggedNodeStatus.UNPROCESSED, result.getStatus());
+    Assertions.assertEquals(LoggedNodeStatus.UNPROCESSED, result.getStatus());
```
Comment (marcinus) [inline handler/core/src/test/java/io/knotx/fragments/handler/consumer/EventLogConverterTest.java]: Is this an intentional change? I think it's a standard to import static these assertions

File: handler/core/src/test/java/io/knotx/fragments/handler/consumer/MetadataConverterTest.java [hunk 5]
```diff
@@ -58,7 +59,7 @@ void shouldProduceEmptyJsonWhenNoMetadataProvided() {
 
     JsonObject output = tested.getExecutionLog().toJson();
 
-    assertJsonEquals(new JsonObject(), output);
+    KnotxAssertions.assertJsonEquals(new JsonObject(), output);
```
Comment (marcinus) [inline handler/core/src/test/java/io/knotx/fragments/handler/consumer/MetadataConverterTest.java]: Is this intentional?

File: handler/core/src/test/java/io/knotx/fragments/handler/consumer/EventLogConverterTest.java (hunk 4 above)
Comment (tomaszmichalak) [reply handler/core/src/test/java/io/knotx/fragments/handler/consumer/EventLogConverterTest.java]: Sure, we should use static imports.

File: handler/core/src/test/java/io/knotx/fragments/handler/consumer/MetadataConverterTest.java (hunk 5 above)
Comment (tomaszmichalak) [reply handler/core/src/test/java/io/knotx/fragments/handler/consumer/MetadataConverterTest.java]: Fixed.

File: handler/core/src/main/java/io/knotx/fragments/handler/consumer/FragmentExecutionLogConsumersNotifier.java (hunk 3 above)
Comment (tomaszmichalak) [reply handler/core/src/main/java/io/knotx/fragments/handler/consumer/FragmentExecutionLogConsumersNotifier.java]: Done

File: README.md [hunk 6]
```diff
@@ -22,6 +22,18 @@ Knot.x Fragments is designed to build fault-tolerant, back-end integrations such
 - API Gateways, Backend For Frontend (BFF) for single-page applications (SPA), Web APIs
 - documents processing (HTML, JSON, PDF etc) with a templating engine support
 
+## Modules
+
+- [Fragments Supplier](https://github.com/Knotx/knotx-fragments/tree/master/supplier) - converts a HTTP request into one or more [**Fragments**](https://github.com/Knotx/knotx-fragments/tree/master/api#knotx-fragment-api)
```
Comment (marcinus) [inline README.md]: Don't know if it's correct, but somehow I feel that `HTTP` should be preceeded with `an`. Call me purist 😆

File: README.md [hunk 7]
```diff
@@ -22,6 +22,18 @@ Knot.x Fragments is designed to build fault-tolerant, back-end integrations such
...
+    - [HTML Splitter](https://github.com/Knotx/knotx-fragments/tree/master/supplier/html-splitter)
+    - [Single Fragment Supplier](https://github.com/Knotx/knotx-fragments/tree/master/supplier/single-fragment)
+- [Fragments Handler](https://github.com/Knotx/knotx-fragments/tree/master/handler) - configures tasks for fragments and delegate processing to Engine
```
Comment (marcinus) [inline README.md]:
```suggestion
- [Fragments Handler](https://github.com/Knotx/knotx-fragments/tree/master/handler) - configures tasks for fragments and delegates processing to Engine 
```

File: README.md [hunk 8]
```diff
@@ -22,6 +22,18 @@ Knot.x Fragments is designed to build fault-tolerant, back-end integrations such
 - API Gateways, Backend For Frontend (BFF) for single-page applications (SPA), Web APIs
 - documents processing (HTML, JSON, PDF etc) with a templating engine support
 
+## Modules
+
+- [Fragments Supplier](https://github.com/Knotx/knotx-fragments/tree/master/supplier) - converts a HTTP request into one or more [**Fragments**](https://github.com/Knotx/knotx-fragments/tree/master/api#knotx-fragment-api)
+    - [HTML Splitter](https://github.com/Knotx/knotx-fragments/tree/master/supplier/html-splitter)
+    - [Single Fragment Supplier](https://github.com/Knotx/knotx-fragments/tree/master/supplier/single-fragment)
+- [Fragments Handler](https://github.com/Knotx/knotx-fragments/tree/master/handler) - configures tasks for fragments and delegate processing to Engine 
+    - [Fragment Execution Log Consumer](https://github.com/Knotx/knotx-fragments/tree/master/handler/consumer) - expose data from evaluated fragments
```
Comment (marcinus) [inline README.md]:
```suggestion
    - [Fragment Execution Log Consumer](https://github.com/Knotx/knotx-fragments/tree/master/handler/consumer) - exposes data from evaluated fragments
```

File: README.md [hunk 9]
```diff
@@ -22,6 +22,18 @@ Knot.x Fragments is designed to build fault-tolerant, back-end integrations such
...
+- [Fragments Engine](https://github.com/Knotx/knotx-fragments/tree/master/engine) - evaluates fragments (tasks) with map-reduce nature
+- [Fragments Assembler](https://github.com/Knotx/knotx-fragments/tree/master/assembler) - merges Fragments into one a single response
```
Comment (marcinus) [inline README.md]:
```suggestion
- [Fragments Assembler](https://github.com/Knotx/knotx-fragments/tree/master/assembler) - merges Fragments into a single response
```

File: handler/consumer/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+listens for fragment processing operations results. When a fragment operation ends, task execution 
+data is exposed to consumers via [Fragment execution log](#what-is-the-execution-log).
+
+## How to configure
+You can choose and configure one of the pre-defined consumers:
+- [Fragment HTML Body Writer](https://github.com/Knotx/knotx-fragments/blob/master/handler/consumer/html)
+- [JSON Writer](https://github.com/Knotx/knotx-fragments/blob/master/handler/consumer/json)
+
+or implement your custom one. Check the [Fragments Handler](https://github.com/Knotx/knotx-fragments/tree/master/handler#how-to-configure) 
+documentation for more configuration details.
```
Comment (marcinus) [inline handler/consumer/README.md]:
```suggestion
documentation for configuration details.
```

File: handler/consumer/README.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+## What is the execution log?
+[Fragment execution log](https://github.com/Knotx/knotx-fragments/blob/master/handler/consumer/api/docs/asciidoc/dataobjects.adoc#fragmentexecutionlog)
+is a JSON that contains both [fragment](https://github.com/Knotx/knotx-fragments/blob/master/api/docs/asciidoc/dataobjects.adoc#fragment) 
+data and [graph node execution log](https://github.com/Knotx/knotx-fragments/blob/master/handler/consumer/api/docs/asciidoc/dataobjects.adoc#graphnodeexecutionlog).
+  
+The [graph node execution log](#graph-node-execution-log) contains data about the processing of the 
+[Task](https://github.com/Knotx/knotx-fragments/tree/master/engine#task) root node and allows you to
```
Comment (marcinus) [inline handler/consumer/README.md]:
```suggestion
[Task](https://github.com/Knotx/knotx-fragments/tree/master/engine#task)'s root node and allows you to 
```

File: handler/consumer/README.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+traverse the entire graph.
+
+### Graph Node Execution log
+[Graph node execution log](https://github.com/Knotx/knotx-fragments/blob/master/handler/consumer/api/docs/asciidoc/dataobjects.adoc#graphnodeexecutionlog) 
+represents a root node execution data. It combines knowledge from sources:
+- [Task definition](https://github.com/Knotx/knotx-fragments/tree/master/engine#task)
+- [Task metadata](https://github.com/Knotx/knotx-fragments/blob/master/handler/api/src/main/java/io/knotx/fragments/handler/api/metadata/TaskMetadata.java)
+- [Fragment's log](https://github.com/Knotx/knotx-fragments/tree/master/engine#fragments-log)
+
+The main difference between the Task definition and graph node execution log are [missing nodes](#missing-nodes).
```
Comment (marcinus) [inline handler/consumer/README.md]:
```suggestion
The main difference between the Task definition and the graph node execution log is presence of [missing nodes](#missing-nodes).
```

File: handler/consumer/README.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) 
+  ends with `_error` transition that is not configured as a node output
+
+  ![Task](assets/images/missing-composite-error.png)
+
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) 
+  ends with `_error` transition and the `_error` node output is configured 
+
+  ![Task](assets/images/missing-composite-custom-with-custom.png)
```
Comment (marcinus) [inline handler/consumer/README.md]: Please double check - seems that image is inconsistent with description

File: handler/consumer/README.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) 
+  ends with a custom transition that is not configured as a node output
+
+  ![Task](assets/images/missing-composite-custom.png)
+
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) ends with custom `fallback` transition and `fallback` transition is configured for the composite node
+
+  ![Task](assets/images/missing-composite-custom-with-error.png)
```
Comment (marcinus) [inline handler/consumer/README.md]: same here

File: handler/consumer/README.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+  ![Task](assets/images/missing-composite-error.png)
+
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) 
+  ends with `_error` transition and the `_error` node output is configured 
+
+  ![Task](assets/images/missing-composite-custom-with-custom.png)
+
+- one of the subgraphs in the [composite node](https://github.com/Knotx/knotx-fragments/tree/master/engine#composite-node) 
+  ends with a custom transition that is not configured as a node output
```
Comment (marcinus) [inline handler/consumer/README.md]: > that is not configured as a node output

which node is this about?

File: handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/FragmentExecutionLogConsumer.java (hunk 2 above)
Comment (marcinus) [reply handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/FragmentExecutionLogConsumer.java]: Fair point. If necessery, we can always `pump up` the interface later.

File: handler/consumer/README.md (hunk 15 above)
Comment (tomaszmichalak) [reply handler/consumer/README.md]: Fixed the comment:
> one of the subgraphs in the composite node ends with the custom fallback transition what indicates that the whole composite ends with _error transition (declared in the task)

**Comment by tomaszmichalak** [comment]
Images fixed. You are right - they were misplaced.

**Pull request merged**